File: 72-edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:

        m=len(word1) + 1
        n=len(word2) + 1
        
        d = {}
        for i in range(m):
            d[i, 0] = i
        
        for j in range(n):
            d[0, j] = j
        
        for j in range(1, n):
            for i in range(1, m):
                cost = 0 if word1[i-1] == word2[j-1] else 1
                d[i, j] = min([
                    d[i-1, j] + 1,       # deletion
                    d[i, j-1] + 1,       # insertion
                    d[i-1, j-1] + cost   # substitution
                ])

        return d[m-1, n-1]

File: 72-edit-distance.cpp
class Solution {
public:
    int32_t minDistance(std::string word1, std::string word2) {
        size_t m = word1.size() + 1;
        size_t n = word2.size() + 1;
        
        std::map<std::pair<size_t, size_t>, size_t> d;
        for (size_t i = 0; i < m; ++i)
        {
            d[std::make_pair(i, 0)] = i;
        }
        
        for (size_t j = 0; j < n; ++j)
        {
            d[std::make_pair(0, j)] = j;
        }
        
        for (size_t j = 1; j < n; ++j)
        {
            for (size_t i = 1; i < m; ++i)
            {
                auto minimum = [](size_t x, size_t y, size_t z) {return std::min(std::min(x, y), z);};
                auto cost = word1.at(i - 1) == word2.at(j - 1) ? 0 : 1;
                d[std::make_pair(i, j)] = minimum(d[std::make_pair(i-1, j)] + 1, d[std::make_pair(i, j-1)] + 1, d[std::make_pair(i-1, j-1)] + cost);
            }
        }
        
        return d[std::make_pair(m-1, n-1)];
    }
};